workflows/mapping/Snakefile

```diff
@@ -249,4 +257,67 @@ rule merge_bam:
     wrapper:
         wrapper_for('samtools/merge')
 
-# vim: ft=python
+
+rule fastq_count:
+    input: '{prefix}.fastq.gz'
+    output: '{prefix}.fastq.count'
+    shell: "gunzip -c {input[0]} | echo $((`wc -l` / 4)) > {output[0]}"
+
+
+rule bam_count:
+    input: '{prefix}.bam'
+    output: '{prefix}.bam.count'
+    shell: "samtools view -c {input[0]} > {output[0]}"
+
+rule sample_count_table:
+    input: fq = '{prefix}.fastq.count',
+           cutadapt = '{prefix}.cutadapt.fastq.count',
+           align = '{prefix}.cutadapt.ht2.bam.count',
+           dedup = '{prefix}.cutadapt.ht2.unique.sort.dedup.bam.count'
+    output: '{prefix}_counts.csv',
+    run:
+        import pandas as pd
+        def read_count(count):
+            with open(count, 'r') as FH:
+                return int(FH.read().strip())
+        
+        fq = read_count(input.fq)
+        cutadapt = read_count(input.cutadapt)
+        align = read_count(input.align)
+        dedup = read_count(input.dedup)
+
+        df = pd.DataFrame([[wildcards.prefix, fq, cutadapt, align, dedup]], columns=['sampleID', 'Number Reads', 'Number Trimmed Reads', 'Number Aligned', 'Number Uniquely Aligned'])
+        df.to_csv(output[0], index=False)
+
+rule count_table:
+    input: SH.build_targets(['{runLevel}_counts.csv'])
+    output: csv = 'count_table.csv',
+            html = 'count_table.html'
+    run:
+        import pandas as pd
+
+        dfs = [pd.read_csv(x) for x in input]
+        df = pd.concat(dfs)
+        df.to_csv(output.csv, index=False)
+        with open(output.html, 'w') as OUT:
+            df.to_html(buf=OUT, index=False, classes='table table-striped')
+
+rule report:
+    input: Table1='count_table.html'
+    output: html='report.html'
+    run:
+        from reporting import report
+
+        myIn = {}
+        with open(input.Table1, 'r') as FH:
+            myIn['Table1'] = FH.read().strip()
+        
```

Comment:
I like it. I think you are saying to create a report specific jinja template. Each template would have all text written and then use jinja to fill in the 'chunks'. Conditional reporting of sections can be done trivially with jinja by wrapping each section in a conditional based on `chunk.id`. This would keep all of the report in one file that could be easily extended 6 months from now. 

How about something like this:

``` python

rule bowtie2_chunk:
    input: _bt2_input_func      # provide list of all *.bam.log
    output: temp('bowtie2_chunk.pkl')
    run:
         import pandas as pd
         import lcdb.odo
         import pickle
         import lcdb.reporting.chunk as chunk
         dfs = []
         for i in input:         
              dfs.append(odo(i))
         df = pd.concat(dfs)
         ck = chunk('bowtie2', df.to_html(classes='table stripped-table')
         pickle.dump(output[0])
```

I don't think the odo syntax is correct, but you get the idea.

I think this is where having the interface module working well would be powerful. Dumping to a pickle would the most powerful because then we can pass python objects to jinja instead of having to load plain text html.